library/pkcs7_write.c

```diff
@@ -0,0 +1,843 @@
+                                bytes_written_in_step, 0 );
+    }
+    if( ret )
+        mbedtls_printf( "ERROR: Failed to add algorithm ID to PKCS7\n" );
+
+    return ( ret );
+}
+
+static int pkcs7_set_version( unsigned char **start, size_t *size,
+                              unsigned char **ptr,
+                              mbedtls_pkcs7_info *pkcs7_info )
+{
+    int ret;
+    /* for now only version 1 */
+    int version = 1;
+
+    ret = pkcs7_set_algo_id( start, size, ptr, pkcs7_info );
+    if( ret != 0 )
+        return ( ret );
+
+    ret = pkcs7_write_data( start, size, ptr, WRITE_VERSION, &version,
+                            sizeof( version ), 0 );
+    if( ret )
+        mbedtls_printf( "ERROR: Failed to add version to PKCS7\n" );
+
+    return ( ret );
+}
+
+static int pkcs7_set_signed_data( unsigned char **start, size_t *size,
+                                  unsigned char **ptr, mbedtls_pkcs7_info *pkcs7_info )
+{
+    int ret;
+    size_t curr_len;
+    unsigned char tag = MBEDTLS_ASN1_CONSTRUCTED | MBEDTLS_ASN1_SEQUENCE;
+
+    ret = pkcs7_set_version( start, size, ptr, pkcs7_info );
+    if( ret != 0 )
+        return ( ret );
+
+    curr_len = *size - ( *ptr - *start );
+    ret = pkcs7_write_data( start, size, ptr,
+                            WRITE_TAG, &tag,
+                            curr_len, 0 );
+    if( ret )
+        mbedtls_printf(
+            "ERROR: Failed to add signed data's SEQUENCE header to PKCS7\n");
+
+    return ( ret );
+}
+
+static int pkcs7_set_pkcs7_oid( unsigned char **start, size_t *size,
+                                unsigned char **ptr,
+                                mbedtls_pkcs7_info *pkcs7_info )
+{
+    int ret;
+    size_t curr_len;
+    unsigned char tag = MBEDTLS_ASN1_CONTEXT_SPECIFIC | MBEDTLS_ASN1_CONSTRUCTED;
+
+    ret = pkcs7_set_signed_data( start, size, ptr, pkcs7_info );
+
+    if( ret != 0 )
+        return ( ret );
+
+    curr_len = *size - ( *ptr - *start );
+    ret = pkcs7_write_data( start, size, ptr,
+                            WRITE_TAG,
+                            &tag, curr_len, 0 );
+    if( ret == 0 ) {
+        curr_len = *size - ( *ptr - *start );
+        ret = pkcs7_write_data(
+                  start, size, ptr, WRITE_OID,
+                  (void *) MBEDTLS_OID_PKCS7_SIGNED_DATA,
+                  strlen( MBEDTLS_OID_PKCS7_SIGNED_DATA ), curr_len );
+    }
+    if( ret )
+        mbedtls_printf(
+            "ERROR: Failed to add PKCS7 OID/SEQUENCE header to PKCS7\n" );
+
+    return ( ret );
+}
+
+static int pkcs7_start_generation( unsigned char **start, size_t *size,
+                                   unsigned char **ptr, mbedtls_pkcs7_info *pkcs7_info )
+{
+    /* This will call every other function before anything is actually written */
+    return ( pkcs7_set_pkcs7_oid( start, size, ptr, pkcs7_info ) );
+}
+
+int mbedtls_pkcs7_create( unsigned char **pkcs7, size_t *pkcs7_size,
```

Comment:
If we use a max size argument then we might as will initially allocate that much memory. At the end of the function, the call to `get_leading_unused_bytes` and `memcpy` will copy the finished buffer into a buffer that is exactly the required length.

Even after writing this, I could not say what a fair estimate of size would be considering the possibility of several signers. This would be a hard thing to put on the user to decide.

It might be easier to macro this out and interally determine a max size as `MAX_SIZE = `PKCS7_EMPTY_SIZE` + (n_signers * MAX_X509_DER_SIZE) + (n_signers * MAX_MD_SIZE)` but things like determining approx tag length get tricky as well.

Whatever you all fee best, I will implement.